ompi/opal/orte/oshmem/test: max hostname length cleanup

kmroz opened the pull request:
Trying to standardize from where max hostname length is derived and how gethostname() called. This is low priority/cleanup, but touches a fair number of files.

I've left MPI_Get_processor_name() to continue using MPI_MAX_PROCESSOR_NAME, as the variable to which it writes the hostname is passed into the function. If for some reason hostname max lengths ever exceed MPI_MAX_PROCESSOR_NAME, we could segfault.

Also, programs in orte/test/ are not configured to build, and further, they have various build failures. I've added the hostname max len fixups to these files as well, knowing they don't build. Perhaps a future TODO :)

Fixes: https://github.com/open-mpi/ompi/issues/1551

hppritcha commented:
@kmroz would you mind rebasing this down to a smaller set of commits?  it helps in tracking when we/if we want to merge in to other branches.  Squashing down to one isn't necessary, more like 2 or 3 would be fine.

kmroz commented:
@hppritcha - Will do.

jsquyres commented on ompi/runtime/ompi_mpi_abort.c:
```diff
@@ -133,7 +133,7 @@ ompi_mpi_abort(struct ompi_communicator_t* comm,
         host = ompi_process_info.nodename;
     } else {
-        gethostname(hostname, sizeof(hostname));
+        gethostname(hostname, MAXHOSTNAMELEN);
```
Is there a reason to make changes like this?  I somewhat like the defensive coding style of using `sizeof(foo)` to indicate the length of an array.  In this case, it's obviously `MAXHOSTNAME`, but using `sizeof` protects you in case the array changes size someday (i.e., there's only 1 place to update instead of 2).

Indeed, I wouldn't mind seeing:

1. Changing the size of all hostname-like arrays to `MAXHOSTNAMELEN` (which you do already in this PR)
1. Changing all references to the length of the array (e.g., passing as the 2nd arg to `gethostname()`) use `sizeof(array_name)`.

jsquyres commented on opal/mca/event/libevent2022/libevent/evdns.c:
```diff
@@ -3108,11 +3106,11 @@ evdns_search_ndots_set(const int ndots) {
 static void
 search_set_from_hostname(struct evdns_base *base) {
-	char hostname[HOST_NAME_MAX + 1], *domainname;
+	char hostname[MAXHOSTNAMELEN], *domainname;
```
FWIW, we import libevent from an external package.  We typically modify as little code in this libevent core code as possible (because then we'll just need to remember to make the same updates the next time we upgrade the internal copy of libevent).

Specifically: it's probably ok / harmless to make these changes as part of this PR, but we should probably expect them to drift back the next time we upgrade libevent.

jsquyres commented on opal/mca/pmix/pmix114/pmix/examples/dynamic.c:
```diff
@@ -45,12 +45,12 @@ int main(int argc, char **argv)
     char nsp2[PMIX_MAX_NSLEN+1];
     pmix_app_t *app;
-    char hostname[1024], dir[1024];
+    char hostname[MAXHOSTNAMELEN], dir[1024];
```
Similar to libevent, the pmix component comes from outside Open MPI, too.  Patches for PMIx should likely go to their upstream, too.

jsquyres commented on opal/util/output.c:
```diff
@@ -176,8 +176,7 @@ bool opal_output_init(void)
     }
-    gethostname(hostname, sizeof(hostname));
-    hostname[sizeof(hostname)-1] = '\0';
+    gethostname(hostname, MAXHOSTNAMELEN);
```
Is this kind of stuff correct?

I ask because, for example, the OS X man page for `gethostname(2)` says:

>      Host names are limited in length to {sysconf(_SC_HOST_NAME_MAX)} characters, not including the trailing null, currently 255.

Since we're using `MAXHOSTNAMELEN` and not `sysconf()`, we can't know that it is big enough to include the terminating character.

kmroz replied:
From the Linux `gethostname(2)` man page:

> SUSv2 guarantees that "Host names are limited to 255 bytes".  POSIX.1 guarantees that "Host names (not including the terminating null  byte) are  limited  to HOST_NAME_MAX bytes".  On Linux, HOST_NAME_MAX is defined with the value 64, which has been the limit since Linux 1.0 (earlier kernels imposed a limit of 8 bytes).

For POSIX systems, I think defining a MAXHOSTNAMELEN = limits.h:HOST_NAME_MAX + 1 and using it in this fashion is fine. We have space for the null terminator which gethostname() will append provided it doesn't overrun. I imagine your concern is for non-POSIX systems (which I also feared a bit).

Perhaps the alternative would be to do:

    +/* From limits.h. Since we're using gethostname(), we add 1 to the max to allow
    + * for null termination.
    + */
    +#ifdef HOST_NAME_MAX
    +#define MAXHOSTNAMELEN HOST_NAME_MAX + 1
    +#else
    +#define MAXHOSTNAMELEN 255 + 1
    +#endif
    #endif

Or just set it to 255 + 1 across the board and not worry about pulling it via limits.h? BTW - what does OS X limits.h:HOST_NAME_MAX say?

jsquyres commented:
@kmroz Pro tip: if you put "Fixes #1551" in a commit message, it'll auto-close that issue when this is merged.  I don't remember offhand, but you *might* get the same effect if you put that string in the original description of this PR...?

kmroz commented:
@jsquyres Ah, cool! I didn't know this. We can test the "PR" hypothesis with this PR... provided I can get it into the right shape :)

bosilca commented on ompi/info/info.c:
```diff
@@ -134,7 +134,7 @@ int ompi_info_init(void)
     /* local host name */
-    gethostname(val, MPI_MAX_INFO_VAL);
+    gethostname(val, MAXHOSTNAMELEN);
     ompi_info_set(&ompi_mpi_info_env.info, "host", val);
```
This is limited by MPI_MAX_INFO_KEY.

kmroz replied:
Thanks! I will look into changing this occurrence.

bosilca commented:
So the plan is to change [where possible] to use sizeof instead of hardcoding the MAXHOSTNAMELEN ?

jsquyres commented:
@bosilca I was specifically referring to using `char foo[X]` to size the arrays (e.g., X = `MAXHOSTNAMELEN`), but then use `sizeof(foo)` when passing the length of the array to something (e.g., `gethostname()`) -- just seems like good defensive programming.

bosilca commented:
That's exactly what I was referring to. I was wondering if the plan was to change all the instances where sizeof(array[]) can be used (and thus provide a consistent usage) or just the one where it has been specifically mentioned ?

kmroz commented:
I will rework this PR to pass sizeof() to all invocations of gethostname().

jsquyres commented:
This PR was just about `MAXHOSTNAMELEN` -- I don't think there's a larger effort to unify the whole code base on `sizeof(foo)`.  But I agree, that would be a Good Thing.  😄

kmroz commented:
Yes of course, I meant using sizeof() in the context of this PR :smile:

bosilca commented:
Much appreciated @kmroz that will put some consistency in the code. @jsquyres I wasn't talking about the entire code base, but at everything touched by this patch.

kmroz commented:
Rebased on top of current master, and s/MAXHOSTNAMELEN/sizeof()/ when invoking gethostname().

Some additional things:

1. In `opal_config_bottom.h` - if `HOST_NAME_MAX` is not defined in `limits.h`, I've increased `MAXHOSTNAMELEN` to the maximum possible value according to SUSv2 (ie. 255 + 1).
2. @bosilca - `ompi/info/info.c:MPI_MAX_INFO_KEY = OPAL_MAX_INFO_VAL = 256`. So, we shouldn't exceed this with `gethostname()` and `MAXHOSTNAMELEN`. Agreed?
3. `pmix_config_bottom.h` - includes `limit.h`, so we should get a valid `HOST_NAME_MAX` on which to base `MAXHOSTNAMELEN`, otherwise I've bumped it to 255 + 1.
4. I've mostly left `get_processor_name.c:MPI_Get_processor_name()` alone and using `MPI_MAX_PROCESSOR_NAME` for now. It's invoked in a number of places with varying allocations for `*name`. I think this would also benefit some cleanup but from another PR as it would touch fortran code, etc... and I've no idea yet what I'm doing there :smile:
5. I've left `win_compat.h` alone. It defines `MAXHOSTNAMELEN` as `_MAX_PATH`. **Any thoughts?**

@jsquyres - If you feel it would be better to drop libevent/pmix modifications from this PR, I'm happy to rebase things without those changes. I could try to go straight upstream, but would need a pointer where upstream is.

Thanks again for all the reviews!

rhc54 commented:
@kmroz Please do a "git diff opal/mca/pmix" and send me the results - I'll upstream them to the pmix master. You can leave them here as well for now and I'll update them with the next PMIx release.

kmroz commented:
@rhc54 - The diff you asked for here: https://gist.github.com/kmroz/e786d5ea9e5dfd90ccc342d97792b997 I'd appreciate your feedback on `pmix_config_bottom.h`. The `MAXHOSTNAMELEN` section appears to only be included as part of `#ifdef MCS_VXWORKS` - which means all the pmix code using MAXHOSTNMELEN is actually getting it from elsewhere :grimacing: I assumed from opal_config_bottom.h, but introducing compiler errors there didn't break pmix. Any ideas where it might be picking up the macro?! Thanks!

kmroz commented:
@rhc54 - I think I narrowed it down. In pmix code, `MAXHOSTNAMELEN` is taken from `/usr/include/asm-generic/param.h`. I've moved the `#ifndef MAXHOSTNAMELEN` higher in `pmix_config_bottom.h` to cover all environments.

Above gist updated with most recent pmix diffset.

rhc54 commented:
Thanks - that indeed looks correct. I'll update this upstream, but let you include it in your commit here to avoid conflicts.

kmroz commented:
@rhc54 - Great, thanks!

jsquyres commented:
@kmroz Sorry for the delay.  Some (very minor) nits:

1. I wonder if we should rename `MAXHOSTNAMELEN` to be `OPAL_MAXHOSTNAMELEN`, given that we are adding one to it, and therefore it is *not* the same as the system `MAXHOSTNAMELEN` (if there is one).  This would reinforce the idea that this value is guaranteed to have enough space for the `\0` terminator.
1. In a small number of places, you multiply the length of the string by `sizeof(char)`.  That probably isn't necessary.

kmroz commented:
@jsquyres - No problem.

I've removed the sizeof(char) instances and updated the PR.
I will investigate s/MAXHOSTNAMELEN/OPAL_MAXHOSTNAMELEN/.

kmroz commented:
@jsquyres - Hmmm, did you have something like this in mind for both `opal_config_bottom.h` and `pmix_config_bottom.h`? Then, globally `s/hostname[MAXHOSTNAMELEN]/hostname[OPAL_MAXHOSTNAMELEN]/`.

    #ifndef MAXHOSTNAMELEN
    /* From limits.h. Since we're using gethostname(), we add 1 to the max to allow
     * for null termination.
     */
    #ifdef HOST_NAME_MAX
    #define OPAL_MAXHOSTNAMELEN HOST_NAME_MAX + 1
    #else
    /* SUSv2 guarantees that "Host names are limited to  255  bytes". */
    #define OPAL_MAXHOSTNAMELEN 255 + 1
    #else
    #define OPAL_MAXHOSTNAMELEN MAXHOSTNAMELEN + 1
    #endif
    #endif

rhc54 commented:
Yes, that is what he is proposing. I would second it as it will avoid future confusion, if you wouldn't mind.

kmroz commented:
@jsquyres / @rhc54 - Hi gents!

OK - I think this should do it. Please have a look and let me know if you think I've missed something.

I've also modified `win_compat.h` to include a `#define OPAL_MAXHOSTNAMELEN MAXHOSTNAMELEN + 1`. Lastly, I've moved the `#define` block in `opal_config_bottom.h` out of the ` #ifdef MCS_VXWORKS`.

@rhc54 - I've updated the pmix diff (https://gist.github.com/kmroz/e786d5ea9e5dfd90ccc342d97792b997) with the latest.

jsquyres commented:
@kmroz Perhaps a little simpler, without the negative logic (pro tip: put a "c" after the triple-single-tick markdown for verbatim, and it syntax highlights for C):

```c
#if defined(MAXHOSTNAMELEN)
#define OPAL_MAXHOSTNAMELEN (MAXHOSTNAMELEN + 1)
#elif defined(HOST_NAME_MAX)
#define OPAL_MAXHOSTNAMELEN (HOST_NAME_MAX + 1)
#else
/* SUSv2 guarantees that "Host names are limited to 255 bytes". */
#define OPAL_MAXHOSTNAMELEN (255 + 1)
#endif
```

Thanks for your patience!  Otherwise, I think it looks great.

kmroz commented:
@jsquyres - On it... will fix it up shortly.

kmroz commented:
@jsquyres - PR updated per your suggestion. Way cleaner, btw :smile: And thanks for the syntax highlighting tip!

@rhc54 - Above gist updated with latest pmix changes.

rhc54 commented:
Ummm...not quite right. The changes in the pmix code base should be PMIX_MAXHOSTNAMELEN, not OPAL_...

Sorry to be a pain.

kmroz commented:
@rhc54 - No pain at all! PR (specfically commit https://github.com/open-mpi/ompi/pull/1552/commits/e1c64e6e59e9e49cab6872bcfe8beae474a73e66) and gist [pmix.diff](https://gist.github.com/kmroz/e786d5ea9e5dfd90ccc342d97792b997) updated.

rhc54 commented:
:+1:

kmroz commented:
@jsquyres - Is there anything else you see that's needed to improve upon this one?

jsquyres commented:
Works for me.  👍

jsquyres commented:
This is probably work PR'ing to the v2.x branch -- probably for v2.0.1.  It's not critical functionality, by any means, but it will help reduce drift between master and the v2.x branch.

kmroz commented:
@jsquyres - Will do, thanks.
